pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -1427,31 +1475,38 @@ func (gce *GCECloud) ExternalID(instance string) (string, error) {
 	}
 
 	// Fallback to GCE API call if metadata server fails to retrieve ID
-	inst, err := gce.getInstanceByName(instance)
+	inst, err := gce.resolveHost(instance)
 	if err != nil {
 		return "", err
 	}
-	return strconv.FormatUint(inst.Id, 10), nil
+	return strconv.FormatUint(inst.ID, 10), nil
 }
 
 // InstanceID returns the cloud provider ID of the specified instance.
 func (gce *GCECloud) InstanceID(instance string) (string, error) {
-	return gce.projectID + "/" + gce.zone + "/" + canonicalizeInstanceName(instance), nil
+	host, err := gce.resolveHost(instance)
+	if err != nil {
+		return "", err
+	}
+	return gce.projectID + "/" + host.Zone + "/" + host.Name, nil
 }
 
 // List is an implementation of Instances.List.
 func (gce *GCECloud) List(filter string) ([]string, error) {
-	listCall := gce.service.Instances.List(gce.projectID, gce.zone)
-	if len(filter) > 0 {
-		listCall = listCall.Filter("name eq " + filter)
-	}
-	res, err := listCall.Do()
-	if err != nil {
-		return nil, err
-	}
 	var instances []string
-	for _, instance := range res.Items {
-		instances = append(instances, instance.Name)
+	// TODO: Parallel?  (But no generics seems to make this harder)
```

Comment:
Agreed that it could be improved, although it's not clear how generics would help you here?  Given that the loop is O(zones) I'm OK with leaving parallelization as a todo.